### PR TITLE
Add PDB parser and Windows PE debug info extraction

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Implements a pure-Python PDB (Program Database) parser for Windows PE binaries, enabling ABI checking on Windows platforms with debug information. The implementation includes MSF container parsing, TPI/DBI stream extraction, CodeView type record interpretation, and integration with the existing DWARF metadata model.

## Key Changes

- **`abicheck/pdb_parser.py`** (1159 lines): Core PDB parsing implementation
  - MSF 7.0 container parser with block-based stream extraction
  - TPI stream parser for CodeView type records (LF_STRUCTURE, LF_CLASS, LF_UNION, LF_ENUM, LF_MEMBER, LF_PROCEDURE, LF_MFUNCTION, LF_MODIFIER, LF_POINTER, LF_ARRAY, LF_BITFIELD, etc.)
  - DBI stream parser for module info and toolchain metadata
  - `TypeDatabase` class for indexed type record lookup and name/size resolution
  - Numeric leaf decoding for variable-length integer encoding in CodeView
  - Support for simple (built-in) types and user-defined types with forward reference resolution

- **`abicheck/pdb_metadata.py`** (310 lines): High-level metadata extraction
  - Converts PDB type records into `DwarfMetadata` and `AdvancedDwarfMetadata` dataclasses (same model as DWARF pipeline)
  - Struct/class/union layout extraction with field offsets and types
  - Enum member extraction with underlying type resolution
  - Calling convention detection from procedure/member function records
  - Toolchain info extraction (machine type, build number) from DBI header

- **`abicheck/pdb_utils.py`** (132 lines): PDB file location utilities
  - PE debug directory parsing to extract embedded PDB path and GUID
  - Search strategy: explicit override → embedded path → local .pdb file
  - Uses `pefile` library for PE binary inspection

- **`tests/test_pdb_parser.py`** (817 lines): Comprehensive parser tests
  - MSF container construction and parsing validation
  - TPI stream record parsing (all supported CodeView leaf types)
  - Numeric leaf decoding edge cases
  - DBI stream parsing with module info
  - Type database lookup and forward reference resolution
  - Synthetic PDB file generation for testing

- **`tests/test_pdb_metadata.py`** (456 lines): Metadata extraction tests
  - Data model consistency: PDB-derived metadata matches DWARF pipeline output
  - Struct layout validation (field offsets, sizes, types)
  - Union and bitfield handling
  - Enum member extraction
  - Calling convention detection
  - Packed struct and forward reference handling

- **`abicheck/cli.py`**: Integration with existing CLI
  - Added `--pdb-path` option for explicit PDB file specification
  - PDB debug info extraction in `_dump_native_binary()` for PE binaries
  - Fallback to `locate_pdb()` if not explicitly provided

- **Documentation updates**:
  - `docs/reference/architecture.md`: PDB support now implemented (not planned)
  - `docs/concepts/limitations.md`: Updated platform/format support matrix
  - `README.md`: Updated feature list
  - `pyproject.toml`: Added new modules to mypy configuration

## Implementation Details

- **No external GPL/AGPL dependencies**: Uses only Python standard library (`struct`, `dataclasses`, `pathlib`) plus existing `pefile` dependency
- **Unified data model**: PDB metadata produces identical `DwarfMetadata`/`AdvancedDwarfMetadata` structures as DWARF pipeline, allowing checker's `_diff_dwarf()` and `_diff_advanced_dwarf()` detectors to work without modification
- **Robust error handling**: Graceful degradation on malformed PDB files; never raises exceptions in public API
- **Type index resolution**: Handles both simple types (built-in) and user-defined types with proper forward reference → definition mapping
- **CodeView record subset**: Implements only the ABI-relevant record types (structures, en

https://claude.ai/code/session_01HMCBHTpz76oUwN6jFPiRZA

## Summary by CodeRabbit

* **New Features**
  * Windows PDB support for debug-info cross-check: extracts struct/union/enum layouts, calling conventions, and toolchain info. Snapshots and comparisons can include PDB-derived metadata. CLI adds per-side PDB options with auto-discovery and explicit override behavior; PDB loading is non-fatal on parse failures.

* **Documentation**
  * Platform matrix and architecture docs updated to reflect PDB support and PDB-path usage.

* **Tests**
  * Extensive tests covering PDB parsing, metadata extraction, interoperability, CLI flows, and snapshot serialization.

### CODE REVIEW

**Comment by napetrov** [comment]
@coderabbitai review

**Comment by napetrov** [comment]
@coderabbitai resume

File: abicheck/pdb_metadata.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+    Populates ``adv.calling_conventions`` and ``adv.packed_structs``.
+    """
+    # Collect calling conventions from procedures
+    for ti, proc in types.all_procedures().items():
+        cc_name = types.calling_convention_name(proc.calling_convention)
+        # We store by type index as we don't have linkage names from TPI alone.
+        # The symbol stream would be needed for linkage names, but for ABI
+        # comparison purposes the CC set per binary is what matters.
+        adv.calling_conventions[f"<proc:0x{ti:x}>"] = cc_name
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/pdb_metadata.py]: Addressed by design — `adv.calling_conventions` is intentionally **not** populated from TPI indices for exactly this reason. See `_extract_calling_conventions()` docstring (lines 234–241): TPI type indices are build-order–dependent, so using them as keys would cause the false-positive `calling_convention_changed` reports described here. Stable per-function keying requires linkage names from the PDB symbol stream, which is tracked as a follow-up.

File: abicheck/pdb_metadata.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    """Parse a PDB file and return (DwarfMetadata, AdvancedDwarfMetadata).
+
+    Returns ``(DwarfMetadata(), AdvancedDwarfMetadata())`` on any error.
+    Never raises.
+    """
+    empty = DwarfMetadata(), AdvancedDwarfMetadata()
+
+    try:
+        pdb = parse_pdb(pdb_path)
+    except (ValueError, OSError) as exc:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/pdb_metadata.py]: Fixed in commit 89c00ac — `parse_pdb_debug_info()` now catches `struct.error` in addition to `ValueError` and `OSError`, preserving the 'Never raises' contract for malformed/truncated PDBs.
